Sprint 17

Sm4k17 opened the pull request.

Sm4k17 committed:
feat: добавлена поддержка английского, немецкого и французского языков

- Добавлены переводы для всего текста в приложении
- Созданы файлы локализации для 4 языков
- Заменены все русские тексты на системные вызовы локализации
- Обновлены все экраны приложения
- Добавлен счетчик дней с поддержкой разных языков

Теперь приложение автоматически показывает интерфейс на языке телефона.

Sm4k17 committed:
git commit -m "Добавил локализацию дат и дней

Что сделано:
- Дата теперь показывается в формате пользователя (русский/английский и т.д.)
- Количество дней правильно склоняется (1 день, 2 дня, 5 дней)
- Поддержка русского, английского, французского и немецкого языков
- Всё автоматически подстраивается под настройки телефона"

Sm4k17 committed:
Локализация приложения Tracker

- Полная миграция строковых литералов на систему локализации
- Добавлена поддержка 4 языков (русский, английский + 2 дополнительных)
- Реализована плюрализация для счетчиков дней через Localizable.stringsdict
- Интеграция R.swift для типобезопасного доступа к ресурсам
- Все пользовательские тексты вынесены в Localizable.strings
- R.swift подключен через Swift Package Plugin (автоматическая генерация)

Sm4k17 committed:
Добавлены snapshot-тесты для всех экранов приложения

Sm4k17 committed:
feat: добавлена темная тема приложения

- Настроены Color Sets для светлого и темного режимов
- Обновлены все snapshot-тесты с поддержкой двух тем
- Цвета автоматически адаптируются под системную тему
- Обновлены скрин-шоты

Sm4k17 committed:
feat: добавить аналитику

- Отслеживание создания трекеров
- Аналитика выполнения/снятия трекеров
- События онбординга и навигации

Sm4k17 committed:
feat: добавлен поиск трекеров

- Реализован поиск трекеров по названию
- Поиск сочетается с фильтрацией по дням недели
- Добавлен плейсхолдер для пустых результатов поиска

Sm4k17 committed:
feat: реализована система фильтрации трекеров

- Добавлена кнопка «Фильтры» на главный экран с корректным позиционированием
- Реализовано модальное окно с 4 вариантами фильтрации:
  • Все трекеры - сброс фильтрации, отображаются все трекеры на выбранную дату
  • Трекеры на сегодня - сброс фильтрации + автоматический переход на текущую дату
  • Завершённые - показ только выполненных трекеров на выбранную дату
  • Незавершённые - показ только невыполненных трекеров на выбранную дату

Ключевые особенности:
- Сохранение состояния фильтра через UserDefaults
- Визуальная индикация активных фильтров (красный текст кнопки)
- Галочки только для фильтров «Завершённые» и «Незавершённые»
- Автоматическое скрытие кнопки фильтра при отсутствии трекеров
- Заглушка «Ничего не найдено» при пустых результатах
- Корректная работа с переключением дат в календаре
- Оптимизированное обновление интерфейса в главном потоке
- Обработка всех крайних случаев и ошибок

Технические детали:
- Использован TrackerFilterService для централизованной фильтрации
- Реализован TrackerFiltersViewController для выбора фильтров
- Добавлены аналитические события для отслеживания использования
- Настроен contentInset для корректного отображения кнопки фильтра

Sm4k17 committed:
feat: реализовано контекстное меню и рефакторинг архитектуры

Что было сделано:

Контекстное меню для трекеров:
- Добавлено контекстное меню по долгому нажатию на ячейку трекера
- Реализованы 3 пункта меню: Закрепить/Открепить, Редактировать, Удалить
- Пункт "Удалить" выделен красным цветом
- Использован современный подход с UIContextMenuInteraction

Редактирование и удаление категорий:
- Добавлено контекстное меню для категорий с 2 пунктами: Редактировать, Удалить
- Реализовано модальное окно для редактирования названия категории
- Добавлен Action Sheet для подтверждения удаления категории
- Валидация уникальности названий и защита от удаления непустых категорий

 Рефакторинг архитектуры:
- Разделен TrackersViewController на несколько компонентов:
  * TrackersDataSource - управление данными и фильтрацией
  * TrackersLayoutCalculator - вычисление layout коллекции
  * TrackersPlaceholderManager - управление состоянием плейсхолдера
- Создан BaseTrackerConfigurationViewController с общей логикой
- EventConfigurationViewController и HabitConfigurationViewController наследуют от базового класса
- Улучшена инкапсуляция и переиспользуемость кода

📱 Улучшения UI/UX:
- Плавные анимации добавления/удаления/обновления элементов
- Валидация ввода и обработка ошибок
- Оптимизирована производительность коллекций
- Добавлена аналитика для отслеживания действий пользователя

Sm4k17 committed:
fix: обновление категорий в коллекции трекеров

- Теперь при изменении названия категории коллекция автоматически обновляется
- TrackerStore слушает изменения в TrackerCategoryStore
- Убраны старые названия категорий в UI

Sm4k17 committed:
## Фикс цвета текста в DatePicker и автоматическое поддержание черной темы

### Проблема:
- Текст в компактном DatePicker постоянно сбрасывал цвет с черного на системный
- При взаимодействии с пикером цвет текста менялся на системный tintColor
- Ручная установка цвета работала только до первого изменения даты

### Решение:
1. **Исследование структуры DatePicker** через дебаггер и вывод иерархии сабвью
2. **Обнаружение динамического UILabel** внутри UIButton, который постоянно переопределяет цвет
3. **Реализация автоматического наблюдателя** с таймером, который каждые 0.1 сек проверяет и исправляет цвет
4. **Добавление KVO** для отслеживания изменений даты с последующей коррекцией цвета

### Технические детали:
- Использован Timer для постоянного мониторинга цвета текста
- Добавлен KVO observer для свойства 'date' DatePicker
- Цвет автоматически исправляется при любом системном изменении
- Фон DatePicker устанавливается ручным способом через цепочку сабвью

### Результат:
- Текст DatePicker всегда остается черным независимо от действий пользователя
- Автоматическое исправление работает в реальном времени
- Убрана необходимость ручной установки цвета при каждом изменении

### Время разработки: ~12 часов

Sm4k17 committed:
feat: добавить аналитику AppMetrica

- Настроена отправка событий во все ключевые экраны
- Отслеживаются открытия/закрытия экранов
- Фиксируются действия: создание трекеров, выполнение, фильтрация, редактирование
- Соответствует требованиям по event/screen/item

Sm4k17 committed:
Рефакторинг DatePicker: замена сложной системы наблюдения на простую установку темы

Sm4k17 committed:
feat: добавить экран статистики с расчетом показателей и empty state

- Реализованы карточки статистики: лучший период, идеальные дни, завершённые трекеры
- Добавлен empty state при отсутствии данных
- Настроена автоматическая пересчёт статистики

Sm4k17 committed:
feat: добавить недостающие ключи локализации для 4 языков

- Добавлены ключи для текстов плейсхолдеров, фильтров и контекстного меню
- Реализована локализация для 'Ничего не найдено', 'Закрепленные', 'Закрепить/Открепить'
- Обновлены экраны статистики, фильтров и контекстных меню
- Поддержка русского, английского, французского и немецкого языков

IlyaDriftman commented on Tracker/Statistic/StatisticsViewController.swift:
```diff
@@ -0,0 +1,209 @@
+        setupUI()
+        setupConstraints()
+        statisticsDataStore.delegate = self
+        statisticsDataStore.refresh()
```
Можно вызывать один раз refresh

IlyaDriftman commented on Tracker/Statistic/StatisticsViewController.swift:
```diff
@@ -0,0 +1,209 @@
+    private let bestPeriodCard = StatisticCard()
+    private let idealDaysCard = StatisticCard()
+    private let completedTrackersCard = StatisticCard()
+    private let averageValueCard = StatisticCard()
```
Можно хранить их в массиве, чтобы избежать повторов: 
private lazy var cards: [StatisticCard] = [
    bestPeriodCard,
    idealDaysCard,
    completedTrackersCard,
    averageValueCard
]
и добавлять их в стек 
cards.forEach { cardsStackView.addArrangedSubview($0) }

IlyaDriftman commented on Tracker/Statistic/StatisticsViewController.swift:
```diff
@@ -0,0 +1,209 @@
+    func statisticsDataStore(_ store: StatisticsDataStore, didUpdate statistics: TrackerStatistics?) {
+        updateUI(with: statistics)
+    }
+}
```
В целом все хорошо, понравилось разделение для удобства чтения, работы и тестирования. Все просто и понятно, читаемо написано

Sm4k17 committed:
feat: добавил аналитику для экрана статистики

Sm4k17 committed:
feat(tracker-config): хедер «N дней» при редактировании + проброс completedDays

- Добавлен хедер с количеством выполненных дней в BaseTrackerConfigurationVC
- Проброс completedDays из TrackersVC → CreationTrackerVC → Habit/Event VC
- Исправлены setupContentStack в дочерних VC (вызов super + наполнение categoryContainer)
- Мелкие правки UI/валидации; аналитика открытия/закрытия экранов

Sm4k17 committed:
force операции

Helvetii951 commented on Tracker/AnalyticsService/AnalyticsService.swift:
```diff
@@ -0,0 +1,34 @@
+    }
+    
+    // Функция для отправки событий
+    func report(event: String, params: [AnyHashable: Any]) {
+        AppMetrica.reportEvent(name: event, parameters: params, onFailure: { error in
+            print("REPORT ERROR: %@", error.localizedDescription)
+        })
+    }
```
Функцию можно сделать статической. Тогда не нужен будет экземпляр `shared`.

Helvetii951 commented on Tracker/Application/AppDelegate.swift:
```diff
@@ -6,14 +6,18 @@
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
-        // Override point for customization after application launch.
+        
+        // Инициализируем все сервисы здесь
+        AnalyticsService.activate() // <- Вся логика здесь
+        
```
Swift - достаточно выразительный язык. Часто можно понять что происходит без комментариев.

Helvetii951 commented on Tracker/Resources/en.lproj/Localizable.strings:
```diff
@@ -0,0 +1,112 @@
+"color" = "Color";
+
+// Messages
+"symbols_limit" = "38 character limit";
+"track_only_what_you_want" = "Track only what you want";
+"even_if_not_water_and_yoga" = "Even if it's not\nliters of water and yoga";
+"cool_technologies" = "That's some cool technology!";
```
Рекомендую называть локализованные строки так:
```
"module_name.screen_name.exact_string_name" = "some string";
```

Или так:

```
"moduleName.screeName.exactStringName" = "some string";

```

Helvetii951 commented on Tracker/Statistic/StatisticCardCell.swift:
```diff
@@ -0,0 +1,102 @@
+        }
+    }
+
+    private let gradientBorderView = GradientBorderView()
+    private let containerView = UIView()
+    private let numberLabel = UILabel()
+    private let descriptionLabel = UILabel()
```
Рекомендую всегда использовать [ленивые свойства](https://www.swiftbysundell.com/articles/using-lazy-properties-in-swift/) для объявления подобного. 

Преимущества:
1. Можно сразу настроить необходимые свойства компонента.
```swift
    private let awesomeButton: UIButton = {
        let button = UIButton()
        button.setTitle("Make some magic", for: .normal)
        button.setTitleColor(.whiteDay, for: .normal)
        button.backgroundColor = .ypGray
        button.layer.cornerRadius = 16
        return button
    }()
```

2. Если какой-то UI компонент/свойство не будет использоваться, можно сэкономить память. 

Подход с ленивыми свойствами используют во многих хороших компаниях.

Helvetii951 commented on Tracker/Statistic/StatisticsViewController.swift:
```diff
@@ -0,0 +1,232 @@
+        emptyStateView.translatesAutoresizingMaskIntoConstraints = false
+        emptyStateView.isHidden = true
+        
+        emptyStateImageView.image = UIImage(named: Constants.placeholderImageName)
```
Надежнее будет воспользоваться генерацией ресурсов xCode 15+:
```swift
    private enum Constants {
        static let placeholderImageName : ImageResource = .statisticsError
        ...
    }
    ...
    emptyStateImageView.image = UIImage(resource: Constants.placeholderImageName)
    ...
    
```
На этапе компиляции мы будем уверены, что картинка есть. А в строке названия можно ошибиться. 

Подробнее можно почитать [здесь](https://sarunw.com/posts/swift-symbols-for-asset-catalog/).

Helvetii951 commented on Tracker/Trackers/TrackerFilters/TrackerFiltersViewController.swift:
```diff
@@ -0,0 +1,126 @@
+        let tableView = UITableView(frame: .zero, style: .insetGrouped)
+        tableView.dataSource = self
+        tableView.delegate = self
+        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "FilterCell")
```
Лучше вынести `reuseIdentifier` в свойство контроллера. Так можно подстраховаться от ошибки неверно введенной строки при регистрации, а затем при переиспользовании.

Helvetii951 commented on TrackerTests/__Snapshots__/TrackerTests/testAddCategoryViewController_iPhone11_Dark.1.png:
Лучше не хранить сгенерированный сторонними библиотеками файлы в репозитории, а выносить в .gitignore.

Helvetii951 commented on Tracker/Trackers/TrackersPlaceholderManager.swift:
```diff
@@ -0,0 +1,54 @@
+        placeholderStackView.isHidden = !isEmpty
+        collectionView.isHidden = isEmpty
+        
+        if isEmpty {
+            if !searchText.isEmpty {
+                placeholderLabel.text = R.string.localizable.nothing_found()
+                placeholderImageView.image = R.image.icSearchEmpty()
+            } else if currentFilter == .completed || currentFilter == .uncompleted {
+                placeholderLabel.text = R.string.localizable.nothing_found()
+                placeholderImageView.image = R.image.icStatsEmpty() ?? R.image.icDizzy()
+            } else {
+                placeholderLabel.text = R.string.localizable.what_to_track()
+                placeholderImageView.image = R.image.icDizzy()
+            }
+        }
```
Часто отдают предпочтение `switch` вместо `else-if`:
```swift
switch (isEmpty, searchText.isEmpty, currentFilter) {
case (true, false, _):
    placeholderLabel.text = R.string.localizable.nothing_found()
    placeholderImageView.image = R.image.icSearchEmpty()
case (true, true, .completed), (true, true, .uncompleted):
    placeholderLabel.text = R.string.localizable.nothing_found()
    placeholderImageView.image = R.image.icStatsEmpty() ?? R.image.icDizzy()
case (true, true, _):
    placeholderLabel.text = R.string.localizable.what_to_track()
    placeholderImageView.image = R.image.icDizzy()
default:
    break
}
```
